### PR TITLE
Split `README` into files located at `/docs`

### LINKED ISSUE #1053: README is absurdly large and unintuitive
Yeah... that's about it. The issue is it causes browsers low-end or even medium-end devices, especially mobile devices, to lag or even completely freeze when trying to load core-js' GitHub repository homepage, or anywhere else that embeds core-js' README, which is a terrible accessibility problem. (and also a little ironic in how a library geared towards supporting old systems fails to do so with the most basic thing like a README)

For a reference benchmark, the current core-js GitHub homepage (https://github.com/zloirock/core-js) will **completely freeze** my browser for roughly a full minute before the README is fully loaded in, with my specs listed below:
- OS: Android 11 aarch64
- CPU: Qualcomm ATOLL-AB (8) @ 1.804GHz
- RAM: 6GB
- Browser¹: Opera 67, Firefox 97, Chrome 98
¹: *Tested in multiple browsers to ensure it isn't a browser-specific issue, although timings were not consistent (Opera freezes for a full minute, Firefox for around 20s, Chrome for around 10s but keeps stuttering upon scrolling), they all froze for some period of time nonetheless.*

As you can see my system a quite decent medium-end system, so I can only fear for those on the lower end. I urge you to consider matching the GitHub README to the [npm README](https://www.npmjs.com/package/core-js), short, concise, and **performant**. The remaining information on the current README (the documentation part, which is 90% of it) should be split up into multiple smaller .md files properly organized in a folder structure inside the `/docs/` folder which already exists on the repository, and only being used for the core-js@3 announcement, which can coexist with the additions. This would make the docs **modular** and **easier to navigate** aswell, so it's a win-win even for those on high-end systems that would never notice this to be a problem. Just imagine if MDN decided to put all their docs into a single monolithic markdown file... even if your PC could handle loading it, it would still suck to traverse.

### PR DESCRIPTION
**Pull request by Rudxain**
# Current plan
Big and "medium" (relatively) sections from the `README.md` will be moved to the `/docs/` directory. Categories will have dedicated inner folders where files will be placed, other files will be placed directly at `docs/` if they are their own "category". To preserve as much external links as possible, the index will be the same (almost, because inner links will be moved to indexes at their respective files) and the sections will have headings that, when clicked, redirect to the corresponding file. This is the reason why the branch name is `readme-redirect`

## Links to more info
https://github.com/zloirock/core-js/pull/1095#issuecomment-1163496767
https://github.com/mdn/content/issues/17561#issuecomment-1164096932
https://github.com/zloirock/core-js/pull/1095#pullrequestreview-1026895683

# Directory tree
This section is about the planned structure of the `/docs` repo tree. This is subject to change and open to suggestions. Indentation = nesting depth. `# ` means it's a heading/section, not a file.
```
usage.md //should this be a folder?
	# Installation
	# `postinstall` message
	# CommonJS API
	# Babel
	# swc
	# Configurable level of aggressiveness
	# Custom build
compatibility-data.md //this includes "Supported engines"
features/
	ecmascript/
		object.md //prefixes seem useless in a folder
		function.md
		error.md
		array.md
		string and regexp.md //is this name good?
		number.md
		math.md
		date.md
		promise.md
		symbol.md
		collections.md
			# Map
			# Set
			# WeakMap
			# WeakSet
		typed-array.md
		reflect.md
		json.md
		global-this.md
	proposals/
		README.md //this contains the index and groups proposals by stage
		//each proposal is its own file, named in the same way as the files in `/packages/core-js/proposals/`
	web-standards/
		structured-clone.md
		base64-utility-methods.md
		set-timeout and set-interval.md
		set-immediate.md
		queue-microtask.md
		url.md
		dom-exception.md
		dom-collections.md
	iteration-helpers.md
missing-polyfills.md
```

# Related Issues
Fixes #1053

# Etc
Other stuff that this PR does is fix some typos and minify URLs

### CODE REVIEW

**Comment by Rudxain** [comment]
Wait, I realized a problem. Any link that links to a nested link will break unless that link has its own dedicated file. I thought it would redirect to the parent, but the link only contains info about the inner entry, not the outer one

**Comment by Rudxain** [comment]
I'll try to preserve as much links as possible, so lots of files will be created. I'll put all ECMAScript files into an `ECMAScript/` folder, and remove all the prefixes because they're redundant

**Comment by zloirock** [comment]
In my vision, on each page should be:
- A kind of header, maybe with a logo, for understanding what it's a part of `core-js` project and, mainly, it should describe what the page contains.
- A full index and links to neighboring pages.

I think that docs should contain a directory tree by categories, at least different directories for usage and polyfills.

I think that pages should contain more than, for example, now I see at `Supported engines.md` - it should be joined with related parts of docs.

As you can see in https://github.com/zloirock/core-js/issues/1053#issuecomment-1057263609, the root `README.md` should still contain all current anchors, but they should contain links to those docs.

**Comment by Rudxain** [comment]
Ok, I'll try doing that. But for now I'll "build the base" first, to make it easier to manage

**Comment by Rudxain** [comment]
I noticed there's lots of absolute URLs in the links. I'll make them relative, to minify files, and make the links "adaptive" to forks of this repo branch

**Comment by zloirock** [comment]
All proposals should be in one folder - when the stage of the proposal changed, a link to the proposal should not be changed.

I'll think about the structure of `usage` and `ecmascript` sections.

**Comment by Rudxain** [comment]
> when the stage of the proposal changed, a link to the proposal should not be changed

The 1st option would be to make each proposal have its own file, but those files are too small, and it would need an extra `README.md` file within the `Proposals/` dir to have an index grouping all proposals by their respective stages.

The 2nd option is to create a `Proposals.md` file and put both the index and all proposals within it. I vote for that option, because it seems easier to navigate and it's more akin to the original root README

**Review by zloirock — COMMENTED** [review]
> The 1st option would be to make each proposal have its own file

It's the only option - the link to the proposal should not be changed after moving between stages, so we can't group them by type. One file for all proposals will be too large - we're trying to get rid of that.

---

`Compatibility data` and `Supported engines` should be joined. Need to rework the structure of `Usage`.

It's better to join all collections (`Set`, `Map`, `WeakSet`, `WeakMap`) on one page and move caveats to this page.

We can join `Number` and `Math` as related.

Links in the index of the root README should be to pages, not to the anchors on the root README.

Too many links now are incorrect or completely broken.

From the legacy anchors in the root README should be removed arrows with links to the index and they should use lower headers.

For features, we should have 3 paths, for consistency with namespaces, now we have similar sections in the README:
- ECMAScript
- ECMAScript proposals
- Web standards

**Comment by Rudxain** [comment]
> We can join `Number` and `Math` as related

I agree, but what if the `BigInt` Math proposal ends in a consensus that favors polymorphism? In that case, `BigInt` would be related too, and also `BigDecimal` (if it ever gets added).

> Links in the index of the root README should be to pages, not to the anchors on the root README.

Definitely agree. Too many indirect redirections are bad

> Too many links now are incorrect or completely broken.

Yep, I noticed, I still need to work on that

About everything else, I agree. I'll start working on it soon after doing something

**Comment by zloirock** [comment]
Please, use kebab-case for filenames for consistency with most files of the repo.

**Comment by Rudxain** [comment]
> Please, use kebab-case for filenames for consistency with most files of the repo.

You mean only files? or all paths? (files & dirs)

**Comment by zloirock** [comment]
I'm for `proposals/function-un-this.md` for [full consistency](https://github.com/zloirock/core-js/blob/master/packages/core-js/proposals/function-un-this.js).

**Comment by Rudxain** [comment]
Now I get it, thank you. So, should I remove the "ECMAScript" prefix?

**Comment by zloirock** [comment]
Yep.

**Comment by zloirock** [comment]
Could you use kebab-case for **all** new files and folders? `/Features/` -> `/features/`, `Missing polyfills.md` -> `missing-polyfills.md`, etc.

**Comment by Rudxain** [comment]
> kebab-case for **all** new files and folders?

Seems like a good idea. I like the consistency

**Comment by Rudxain** [comment]
@zloirock , what do I do with ["`string` and `regexp`"](https://github.com/Rudxain/core-js/blob/readme-redirect/docs/features/ecmascript/string%20and%20regexp.md), ["`set-timeout` and `set-interval`"](https://github.com/Rudxain/core-js/blob/readme-redirect/docs/features/web-standards/set-timeout%20and%20set-interval.md), and ["Base64 utility methods"](https://github.com/Rudxain/core-js/blob/readme-redirect/docs/features/web-standards/Base64%20utility%20methods.md) ?

**Comment by zloirock** [comment]
@Rudxain I think, the same - kebab-case. For example, `base64-utility-methods.md`. Maybe it makes sense to come up with alternate names for some categories, for example, `text-processing` for "`String` and `RegExp`", however at this moment let's leave it as is.

**Comment by Rudxain** [comment]
I'll convert to non-draft because the PR seems to be almost ready

File: README.md [hunk 1]
```diff
@@ -28,3326 +28,235 @@
...
+###### [`Symbol.prototype.description`](docs/features/proposals/symbol-description.md)
+###### [Well-formed `JSON.stringify`](docs/features/proposals/well-formed-stringify.md)
+
+##### [Stage 3 proposals](docs/features/proposals/README.md#stage-3)
+###### [`Array` grouping](docs/features/proposals/array-grouping.md)
+###### [Change `Array` by copy](docs/features/proposals/change-array-by-copy.md)
+
+##### [Stage 2 proposals](docs/features/proposals/README.md#stage-2)
+###### [Iterator helpers](docs/features/proposals/iterator-helpers.md)
+####### [Caveats](docs/features/proposals/iterator-helpers.md#caveats)
```
Comment (zloirock) [inline README.md:210]: image

Too low header.

image

And is not readable below - it's better to use something else, for example, lines, for split sections.

File: docs/example.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import Promise from 'core-js-pure/actual/promise';
+import structuredClone from 'core-js-pure/actual/structured-clone';
+import queueMicrotask from 'core-js-pure/actual/queue-microtask';
+
+from(new Set([1, 2, 3, 2, 1]));                // => [1, 2, 3]
+group([1, 2, 3, 4, 5], it => it % 2);          // => { 1: [1, 3, 5], 0: [2, 4] }
+Promise.resolve(42).then(x => console.log(x)); // => 42
+structuredClone(new Set([1, 2, 3]));           // => new Set([1, 2, 3])
+queueMicrotask(() => console.log('called as microtask'));
+```
```
Comment (zloirock) [inline docs/example.md:42]: It's better to leave it on the main README.

File: docs/features/proposals/README.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+## [`String.prototype.trimStart` / `String.prototype.trimEnd`](string-left-right-trim.md)
+## [`RegExp` `s` (`dotAll`) flag](regexp-dotall-flag.md)
+## [`RegExp` named capture groups](regexp-named-groups.md)
+## [`Promise.allSettled`](promise-all-settled.md)
+## [`Promise.any`](promise-any.md)
+## [`Promise.prototype.finally`](promise-finally.md)
+## [`Symbol.asyncIterator` for asynchronous iteration](async-iteration.md)
+## [`Symbol.prototype.description`](symbol-description.md)
+## [Well-formed `JSON.stringify`](well-formed-stringify.md)
+
```
Comment (zloirock) [inline docs/features/proposals/README.md:35]: In cases like this, it's better to use low headers, now it's not readable.

File: docs/features/proposals/collection-of-from.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  static from(iterable: Iterable<mixed>, mapFn?: (value: any, index: number) => any, thisArg?: any): Set;
+  addAll(...args: Array<mixed>): this;
+  deleteAll(...args: Array<mixed>): boolean;
+  every(callbackfn: (value: any, key: any, target: any) => boolean, thisArg?: any): boolean;
+  filter(callbackfn: (value: any, key: any, target: any) => boolean, thisArg?: any): Set;
+  find(callbackfn: (value: any, key: any, target: any) => boolean), thisArg?: any): any;
+  join(separator: string = ','): string;
+  map(callbackfn: (value: any, key: any, target: any) => any, thisArg?: any): Set;
+  reduce(callbackfn: (memo: any, value: any, key: any, target: any) => any, initialValue?: any): any;
+  some(callbackfn: (value: any, key: any, target: any) => boolean, thisArg?: any): boolean;
```
Comment (zloirock) [inline docs/features/proposals/collection-of-from.md:15]: They should be in `collection-methods`.

File: docs/example.md (hunk 2 above)
Comment (Rudxain) [reply docs/example.md:42]: I agree that it should be on the root-main README for easy/fast access, but it seems kinda big. What if we split the examples in 2 parts? (one in root, the other in docs)

Comment (Rudxain) [reply docs/example.md:42]: I'm going to put it back at root. Maybe in the future (if more examples are added) it would be good to consider splitting it

**Comment by Rudxain** [comment]
I forgot that [this file](https://github.com/Rudxain/core-js/blob/readme-redirect/docs/features/ecmascript/README.md) was still unfinished. Sorry. I'll finish it now

**Comment by zloirock** [comment]
@Rudxain I'm sorry that tris PR still is not merged. It's not because I'm ignoring it or have no time to review it. Looking at the result, I'm still not sure that now it's better than the current monolith - users can be confused or get lost. I'm thinking about adding proper indexes / navigation to it, but I'm still unsure about the best approach. Maybe even a statical site generation from markdown could be better.

**Comment by Rudxain** [comment]
> I'm sorry that tris PR still is not merged. It's not because I'm ignoring it or have no time to review it.

It's fine, don't worry.

> Looking at the result, I'm still not sure that now it's better than the current monolith - users can be confused or get lost. I'm thinking about adding proper indexes / navigation to it, but I'm still unsure about the best approach. Maybe even a statical site generation from markdown could be better.

I partly agree. Maybe we should use GH-Pages instead? Other repos use that approach. It's more flexible because we can add a nav-bar and a burger-menu

**Comment by zloirock** [comment]
Sure, in the case of static site generation most likely will be used GH-Pages - and now [it's already used for `core-js-compat` tests](http://zloirock.github.io/core-js/compat/) (but without markdown). If you mean Jekyll - it's one of the options which I am considering.

**Comment by Jimimaku** [comment]
> Ok, I'll try doing that. But for now I'll "build the base" first, to make it easier to manage

**Comment by Jimimaku** [comment]
> I noticed there's lots of absolute URLs in the links. I'll make them relative, to minify files, and make the links "adaptive" to forks of this repo branch

File: docs/features/proposals/collection-of-from.md (hunk 4 above)
Comment (Santiagosaiz) [reply docs/features/proposals/collection-of-from.md]: 19f2464e9ef9149eef13fc9b02fa3e3951b34661

Comment (Rudxain) [reply docs/features/proposals/collection-of-from.md]: @Santiagosaiz I don't understand. Why is that commit related to this?

**Comment by Rudxain** [comment]
Superseded by #1221

I won't delete `readme-redirect`, until #1221 is merged. This is an exception to [my rule of deleting branches and forks ASAP](https://github.com/Rudxain#repos)

**Comment by Rudxain** [comment]
Superseded by #1221

I won't delete `readme-redirect`, until #1221 is merged. This is an exception to [my rule of deleting branches and forks ASAP](https://github.com/Rudxain/Rudxain/blob/6efa08367a63135e32f391cba1aa3634d96ce310/README.md?plain=1#L35)

**Comment by Rudxain** [comment]
I'll delete the branch as it's very outdated, and can be recreated from this PR

**Comment by zloirock** [comment]
Thanks for working on this!

I wanted to do this right along with the website release. Since it was quite a large amount of work, the opportunity to work on this only arose a few months ago.

You could find the current result here: https://core-js.io/

**Comment by Rudxain** [comment]
You're welcome! I'm happy to be one of the people to make that possible (even though this PR seems like a small fraction of the end product :sweat_smile:)

**Pull request closed without merging**